hydroflow_lang/src/graph/mod.rs

```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
     }
 }
 
-impl From<FlatGraph> for PartitionedGraph {
-    fn from(mut flat_graph: FlatGraph) -> Self {
-        // Algorithm:
-        // 1. Each node begins as its own subgraph.
-        // 2. Collect edges. Sort so edges which should not be split across a handoff come first.
-        // 3. For each edge, try to join `(to, from)` into the same subgraph.
+type AdjList = SecondaryMap<GraphNodeId, BTreeMap<IndexInt, (GraphNodeId, IndexInt)>>;
+
+fn find_subgraphs(
+    nodes: &mut SlotMap<GraphNodeId, Node>,
+    preds: &mut AdjList,
+    succs: &mut AdjList,
+    barrier_crossers: &[(GraphNodeId, GraphNodeId, IndexInt, InputBarrier)],
+) -> (
+    AdjList,
+    AdjList,
+    SecondaryMap<GraphNodeId, GraphSubgraphId>,
+    SlotMap<GraphSubgraphId, Vec<GraphNodeId>>,
+) {
+    // Algorithm:
+    // 1. Each node begins as its own subgraph.
+    // 2. Collect edges. (Future optimization: sort so edges which should not be split across a handoff come first).
+    // 3. For each edge, try to join `(to, from)` into the same subgraph.
+
+    // Pre-calculate node colors.
+    let mut node_color: SecondaryMap<GraphNodeId, Option<Color>> = nodes
+        .keys()
+        .map(|node_id| {
+            let inn_degree = preds[node_id].len();
+            let out_degree = succs[node_id].len();
+            let op_color = node_color(&nodes[node_id], inn_degree, out_degree);
+            (node_id, op_color)
+        })
+        .collect();
+    let mut node_union: UnionFind<GraphNodeId> = UnionFind::with_capacity(nodes.len());
+    // All edges which belong to a single subgraph. Other & self-edges become handoffs.
+    let mut subgraph_edges: HashSet<(EdgePortRef, EdgePortRef)> = Default::default();
+
+    // Sort edges here (for now, no sort/priority).
+    loop {
+        let mut updated = false;
+        for ((src, src_idx), (dst, dst_idx)) in iter_edges(succs) {
+            // Ignore new self-loops.
```

Comment:
Comment why we ignore -- is it because they already have been unioned? If so should maybe add an assertion to that effect?